espnet2/asr/frontend/s3prl.py

```diff
@@ -36,75 +21,37 @@ def __init__(
+        if fs != 16000:
+            logging.warning("All the upstream models in S3PRL now only support 16 kHz audio.")
 
+        # TODO(simpleoier): to remove download_dir and update related configs.
         if download_dir is not None:
             torch.hub.set_dir(download_dir)
 
+        assert frontend_conf.get("upstream", None) in S3PRLUpstream.available_names()
+        upstream = S3PRLUpstream(frontend_conf.get("upstream"))
+        upstream.eval()
+        featurizer = Featurizer(upstream)
+
         self.multilayer_feature = multilayer_feature
-        self.upstream, self.featurizer = self._get_upstream(frontend_conf)
+        self.upstream, self.featurizer = upstream, featurizer
         self.pretrained_params = copy.deepcopy(self.upstream.state_dict())
-        self.output_dim = self.featurizer.output_dim
         self.frontend_type = "s3prl"
-        self.hop_length = self.upstream.get_downsample_rates("key")
-
-    def _get_upstream(self, frontend_conf):
-        """Get S3PRL upstream model."""
-        s3prl_args = base_s3prl_setup(
-            Namespace(**frontend_conf, device="cpu"),
-        )
-        self.args = s3prl_args
-
-        try:
-            import s3prl  # noqa
-        except ModuleNotFoundError:
-            raise ModuleNotFoundError(
-                "s3prl is not installed, please git clone s3prl"
-                " (DO NOT USE PIP or CONDA) "
-                "and install it from Github repo, "
-                "by cloning it locally."
-            )
-        s3prl_path = Path(os.path.abspath(s3prl.__file__)).parent.parent
-        if not os.path.exists(os.path.join(s3prl_path, "hubconf.py")):
-            raise RuntimeError(
-                "You probably have s3prl installed as a pip"
-                "package, please uninstall it and then install it from "
-                "the GitHub repo, by cloning it locally."
-            )
-
-        s3prl_upstream = torch.hub.load(
-            s3prl_path,
-            s3prl_args.upstream,
-            ckpt=s3prl_args.upstream_ckpt,
-            model_config=s3prl_args.upstream_model_config,
-            refresh=s3prl_args.upstream_refresh,
-            source="local",
-        ).to("cpu")
-
-        if getattr(
-            s3prl_upstream, "model", None
-        ) is not None and s3prl_upstream.model.__class__.__name__ in [
-            "Wav2Vec2Model",
-            "HubertModel",
-        ]:
-            s3prl_upstream.model.encoder.layerdrop = 0.0
-
-        from s3prl.upstream.interfaces import Featurizer
-
-        if self.multilayer_feature:
-            feature_selection = "hidden_states"
-        else:
-            feature_selection = "last_hidden_state"
-        s3prl_featurizer = Featurizer(
-            upstream=s3prl_upstream,
-            feature_selection=feature_selection,
-            upstream_device="cpu",
-        )
-
-        return s3prl_upstream, s3prl_featurizer
+        self.hop_length = self.upstream.downsample_rates
```

Comment:
This `self.upstream.downsample_rates` return a list (number of layers) of downsampling rates (e.g. [160, 160, 160, ...]) to explicitly defining every layer's downsampling rate. Also, `self.upstream.hidden_sizes` is similar (e.g. [768, 768, 768, ...]). This is designed to cope with future upstream with different internal strides or hidden sizes. Currently all upstream have the same stride/hidden_size for all layers, so you can simply take the first one.

Reply:
Since Featurizer is used to reduce multiple hidden states into a single hidden state (single stride and hidden size), how about I add an `downsample_rate` attribute in Featurizer, so you know the dimension and the stride after the reduction?

Reply:
Also, would it be better to explicitly make this attribute an public API like `self.hop_length()`? (to align with `self.output_size()`). This is just a random thought, feel free to adopt it or not.

Reply:
Thanks for letting me know this. I think it is good to add downsample_rate in s3prl featurizer. Also, this hop_length variable seems not be used. I don't know why we defined here before. I'm trying to figure that out.

Reply:
It seems to appear in `fused.py`, but I only went through that file very briefly.

Reply:
Cool I am updating this new property to S3PRL master